### PR TITLE
[FLINK-12737][table-api][table-planner] Loosen Table class dependencies

### PR DESCRIPTION
**Pull request by dawidwys**
## What is the purpose of the change

It reduces the dependencies of `TableImpl` that block us from moving it to the api module.

## Brief change log

See commitss

## Verifying this change

This change is already covered by existing tests.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (**yes** / no)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)
  - If yes, how is the feature documented? (**not applicable** / docs / JavaDocs / not documented)

### CODE REVIEW

**Comment by zentol** [comment]
please update the PR title

**Comment by dawidwys** [comment]
Thank you @zentol for reminding!

**Comment by dawidwys** [comment]
@flinkbot attention @twalthr @aljoscha

File: flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/functions/TemporalTableFunction.java [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+import org.apache.flink.table.operations.TableOperation;
+import org.apache.flink.types.Row;
+
+import java.sql.Timestamp;
+
+/**
+ * Class representing temporal table function over some history table. A
+ * {@link TemporalTableFunction} is also an instance of {@link TableFunction}.
+ *
+ * <p>Currently {@link TemporalTableFunction}s are only supported in streaming.
```
Comment (twalthr) [inline flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/functions/TemporalTableFunction.java]: nit: `Currently,`

File: flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/functions/TemporalTableFunction.java [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+	}
+
+	public TableOperation getUnderlyingHistoryTable() {
+		if (underlyingHistoryTable == null) {
+			throw new IllegalStateException("Accessing table field after planing/serialization");
+		}
+		return underlyingHistoryTable;
+	}
+
+	public static TemporalTableFunction create(
```
Comment (twalthr) [inline flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/functions/TemporalTableFunction.java]: can we make this package private or avoid that users use this method? The least we can do is marking it `@Internal` but this is still not perfect. The `Impl` approach was nicer.

File: flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/functions/TemporalTableFunction.java [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+import java.sql.Timestamp;
+
+/**
+ * Class representing temporal table function over some history table. A
+ * {@link TemporalTableFunction} is also an instance of {@link TableFunction}.
+ *
+ * <p>Currently {@link TemporalTableFunction}s are only supported in streaming.
+ */
+@PublicEvolving
+public class TemporalTableFunction extends TableFunction<Row> {
```
Comment (twalthr) [inline flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/functions/TemporalTableFunction.java]: `final`

File: flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/functions/TemporalTableFunction.java (hunk 2 above)
Comment (dawidwys) [reply flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/functions/TemporalTableFunction.java]: We can still go with the `Impl` approach. Will update it.

File: flink-table/flink-table-planner/src/test/scala/org/apache/flink/table/plan/TimeIndicatorConversionTest.scala [hunk 4]
```diff
@@ -253,9 +253,9 @@ class TimeIndicatorConversionTest extends TableTestBase {
         ),
         term("groupBy", "long"),
         term("window", "TumblingGroupWindow('w2, 'newrowtime, 1000.millis)"),
-        term("select", "long", "SUM(int) AS TMP_3", "end('w2) AS TMP_2")
+        term("select", "long", "SUM(int) AS TMP_1", "end('w2) AS TMP_0")
```
Comment (twalthr) [inline flink-table/flink-table-planner/src/test/scala/org/apache/flink/table/plan/TimeIndicatorConversionTest.scala]: side comment: If we would `TMP_` to `EXPR$`, I think we would get closer to Calcite's output.

**Pull request merged**